### PR TITLE
Shape iterator improvements

### LINKED ISSUE #1231: Orderd edge iteration
For some applications it is desired to iterate over edges in their connection order. See also #1169 

Links to OCCT documentation:
*  [wire explorer](https://dev.opencascade.org/doc/refman/html/class_b_rep_tools___wire_explorer.html)

### PR DESCRIPTION
**Pull request by adam-urbanczyk**
Closes #1231 and  #565 

- [x] adds an `__iter__` method to Shape (specialized for Wire)
- [x] ancestors iterator
- [x] siblings iterator
- [x] ancestors/siblings for workplane
- [x] tests
- [x] docs

### CODE REVIEW

**Comment by adam-urbanczyk** [comment]
@jmwright @lorenzncode this is becoming quite big, so could you take a look and share what you think?

Initially I thought that ancestors should somehow become selectors, but the semantics of it would be super confusing (e.g.  `Faces('ancestors')` would return something else than faces). I'm also not sure if the naming of the methods is optimal.

**Comment by lorenzncode** [comment]
> Last, but not least, I think I improved _collectProperty - no more direct hashCode calls. This should the last source of flaky tests.

Excellent, that's good news.

I started playing around with this.  Perhaps change `siblings` to eliminate the `kind` parameter or make it optional?  Say I am selecting edges:

```py
import cadquery as cq

w = cq.Workplane().box(1,1,1)

r0 = w.faces('>Z').edges('>X').siblings('Shell') # empty - invalid kind
r1 = w.faces('>Z').edges('>X').siblings('Face')  # empty - invalid kind
r2 = w.faces('>Z').edges('>X').siblings('Edge')  # empty - invalid kind
r3 = w.faces('>Z').edges('>X').siblings('Vertex')  # 4 edges
```

I think any kind other than 'Vertex' is invalid because the kind should be of inferior type.  Automatically set kind based on the object type for ease of use?
Shell->Face
Face->Edge
Edge->Vertex

**Comment by adam-urbanczyk** [comment]
> I started playing around with this. Perhaps change `siblings` to eliminate the `kind` parameter or make it optional? Say I am selecting edges:
> 
> ```python
> import cadquery as cq
> 
> w = cq.Workplane().box(1,1,1)
> 
> r0 = w.faces('>Z').edges('>X').siblings('Shell') # empty - invalid kind
> r1 = w.faces('>Z').edges('>X').siblings('Face')  # empty - invalid kind
> r2 = w.faces('>Z').edges('>X').siblings('Edge')  # empty - invalid kind
> r3 = w.faces('>Z').edges('>X').siblings('Vertex')  # 4 edges
> ```
> 
> I think any kind other than 'Vertex' is invalid because the kind should be of inferior type. Automatically set kind based on the object type for ease of use? Shell->Face Face->Edge Edge->Vertex

I was thinking about this, but it is not obvious that you always want connectivity via the -1 level. As as user you should consciously choose based on what you want.

File: doc/selectors.rst [hunk 1]
```diff
@@ -142,4 +142,31 @@ It is possible to use user defined vectors as a basis for the selectors. For exa
     result = cq.Workplane("XY").box(10, 10, 10)
 
     # chamfer only one edge
-    result = result.edges(">(-1, 1, 0)").chamfer(1)
+    result = result.edges('>(-1, 1, 0)').chamfer(1)
```
Comment (jmwright) [inline doc/selectors.rst:145]: I'm surprised that black isn't complaining about the single quotes here.

**Review by jmwright — APPROVED** [review]
Looks good, thanks @adam-urbanczyk !

File: doc/selectors.rst (hunk 1 above)
Comment (lorenzncode) [reply doc/selectors.rst]: My understanding was PR #1220 was a manual black run (or blacken-docs).  It is not automated.

File: cadquery/cq.py [hunk 2]
```diff
@@ -1046,6 +1051,43 @@ def compounds(
         """
         return self._selectObjects("Compounds", selector, tag)
 
+    def ancestors(self: T, kind: Shapes, tag: Optional[str] = None) -> T:
+        """
+        Iterate over ancestors.
+
+        :param kind: kind of ancestor, e.g. "Face" or "Edge"
+        :param tag: if set, search the tagged object instead of self
+        :return: a Workplane object who's stack contains all ancestors.
```
Comment (lorenzncode) [inline cadquery/cq.py:1060]: I'll do some testing today.  First a couple of docstring grammar/typo issues.

"who's" should be "whose".  I see this form comes from other Workplane docstrings.

```
$ find ./cadquery -name "*.py" -exec grep -Hn "who's" {} \;
./cadquery/cq.py:865:        :return: a CQ object who's stack contains  the *distinct* vertices of *all* objects on the
./cadquery/cq.py:899:        :return: a CQ object who's stack contains all of the *distinct* faces of *all* objects on
./cadquery/cq.py:934:        :return: a CQ object who's stack contains all of the *distinct* edges of *all* objects on
./cadquery/cq.py:968:        :return: a CQ object who's stack contains all of the *distinct* wires of *all* objects on
./cadquery/cq.py:994:        :return: a CQ object who's stack contains all of the *distinct* solids of *all* objects on
./cadquery/cq.py:1023:        :return: a CQ object who's stack contains all of the *distinct* shells of *all* objects on
./cadquery/cq.py:1046:        :return: a CQ object who's stack contains all of the *distinct* compounds of *all* objects on
./cadquery/cq.py:1060:        :return: a Workplane object who's stack contains all ancestors.
./cadquery/cq.py:1080:        :return: a Workplane object who's stack contains all ancestors.
```

File: cadquery/occ_impl/exporters/dxf.py [hunk 3]
```diff
@@ -173,6 +176,21 @@ def add_shape(self, workplane: Workplane, layer: str = "") -> Self:
 
         return self
 
+    @staticmethod
+    def _ordered_edges(s: Shape) -> List[Edge]:
+
+        rv: List[Edge] = []
+
+        # iterato over wires and then edges
```
Comment (lorenzncode) [inline cadquery/occ_impl/exporters/dxf.py:184]: typo iterate

File: cadquery/cq.py [hunk 4]
```diff
@@ -1046,6 +1051,43 @@ def compounds(
...
+        return self.newObject(set(el for res in results for el in res))
+
+    def siblings(self: T, kind: Shapes, level: int = 1, tag: Optional[str] = None) -> T:
+        """
+        Iterate over siblings.
+
+        :param kind: kind of linking element, e.g. "Vertex" or "Edge"
+        :param level: level of relation - how many elements of kind are in the link
+        :param tag: if set, search the tagged object instead of self
+        :return: a Workplane object who's stack contains all ancestors.
```
Comment (lorenzncode) [inline cadquery/cq.py:1080]: change "ancestors" to "siblings".

Change the return value docstring to something like?:
```        :return: a Workplane object whose stack contains the selected siblings.```

or ```... contains the siblings.```

Comment (adam-urbanczyk) [reply cadquery/cq.py]: Fixed

File: cadquery/occ_impl/exporters/dxf.py (hunk 3 above)
Comment (adam-urbanczyk) [reply cadquery/occ_impl/exporters/dxf.py]: Fixed

File: cadquery/cq.py (hunk 2 above)
Comment (adam-urbanczyk) [reply cadquery/cq.py]: Fixed

**Review by lorenzncode — APPROVED** [review]
It looks like the iterator improvements and new selectors will be useful features!  Thanks @adam-urbanczyk!

Currently `siblings` supports a single level value. I don't have a real use case in mind for this but was thinking about whether it would be useful to allow multiple levels to be returned in a single call.  Say I want levels 1,2 or (2,3,4 excluding 1).  Perhaps a future enhancement if there was a real need.

**Pull request merged**